funnel/models/post.py

```diff
@@ -0,0 +1,260 @@
+    )
+    state = StateManager('_state', POST_STATE, doc="Post state")
+
+    user_id = db.Column(None, db.ForeignKey('user.id'), nullable=False, index=True)
+    user = with_roles(
+        db.relationship(
+            User, backref=db.backref('posts', lazy='dynamic'), foreign_keys=[user_id],
+        ),
+        grants={'author'},
+    )
+
+    profile_id = db.Column(None, db.ForeignKey('profile.id'), nullable=True, index=True)
+    profile = with_roles(
+        db.relationship(Profile, backref=db.backref('posts', lazy='dynamic'),),
+        grants_via={None: {'admin': 'profile_admin'}},
+    )
+
+    project_id = db.Column(None, db.ForeignKey('project.id'), nullable=True, index=True)
+    project = with_roles(
+        db.relationship(Project, backref=db.backref('posts', lazy='dynamic'),),
+        grants_via={None: {'editor': 'project_editor'}},
+    )
+
+    body = MarkdownColumn('body', nullable=False)
+
+    #: Like pinned tweets. You can keep posting updates,
+    #: but might want to pin an update from a week ago.
+    pinned = db.Column(db.Boolean, default=False, nullable=False)
+
+    published_by_id = db.Column(
+        None, db.ForeignKey('user.id'), nullable=True, index=True
+    )
+    published_by = db.relationship(
+        User,
+        backref=db.backref('published_posts', lazy='dynamic'),
+        foreign_keys=[published_by_id],
+    )
+    published_at = db.Column(db.TIMESTAMP(timezone=True), nullable=True)
+
+    deleted_by_id = db.Column(None, db.ForeignKey('user.id'), nullable=True, index=True)
+    deleted_by = db.relationship(
+        User,
+        backref=db.backref('deleted_posts', lazy='dynamic'),
+        foreign_keys=[deleted_by_id],
+    )
+    deleted_at = db.Column(db.TIMESTAMP(timezone=True), nullable=True)
+
+    edited_at = db.Column(db.TIMESTAMP(timezone=True), nullable=True)
+
+    voteset_id = db.Column(None, db.ForeignKey('voteset.id'), nullable=False)
+    voteset = db.relationship(Voteset, uselist=False)
+
+    commentset_id = db.Column(None, db.ForeignKey('commentset.id'), nullable=False)
+    commentset = db.relationship(
+        Commentset,
+        uselist=False,
+        lazy='joined',
+        cascade='all',
+        single_parent=True,
+        backref=db.backref('post', uselist=False),
+    )
+
+    search_vector = db.deferred(
+        db.Column(
+            TSVectorType(
+                'name',
+                'title',
+                'body_text',
+                weights={'name': 'A', 'title': 'A', 'body_text': 'B'},
+                regconfig='english',
+                hltext=lambda: db.func.concat_ws(' / ', Post.title, Post.body_html,),
+            ),
+            nullable=False,
+        )
+    )
+
+    __table_args__ = (
+        db.CheckConstraint(
+            db.case([(profile_id.isnot(None), 1)], else_=0)
+            + db.case([(project_id.isnot(None), 1)], else_=0)
+            == 1,
+            name='post_owner_check',
+        ),
+    )
+
+    __roles__ = {
```

Comment:
`name` and `title` should be in `all`. We will want to enumerate restricted posts for all users, to show them what they're missing. Content will be marked participants-only.